tools/topology/topology1/sof-adl-nau8825.m4

```diff
@@ -0,0 +1,362 @@
+DAI_ADD(sof/pipe-dai-playback.m4,
+        2, SSP, 0, SSP0-Codec,
+        PIPELINE_SOURCE_2, 2, s32le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# capture DAI is SSP0 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+        3, SSP, 0, SSP0-Codec,
+        PIPELINE_SINK_3, 2, s32le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp1 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	5, HDA, 0, iDisp1,
+	PIPELINE_SOURCE_5, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp2 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	6, HDA, 1, iDisp2,
+	PIPELINE_SOURCE_6, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp3 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	7, HDA, 2, iDisp3,
+	PIPELINE_SOURCE_7, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# playback DAI is iDisp4 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	8, HDA, 3, iDisp4,
+	PIPELINE_SOURCE_8, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+ifdef(`NO_AMP',,`
+ifdef(`AMP_SSP',`',`fatal_error(note: Define AMP_SSP for speaker amp SSP Index)')
+ifdef(`SMART_AMP',,`
+# playback DAI is SSP1 using 2 periods
+# Buffers use s16le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+        1, SSP, SPK_SSP_INDEX, SPK_SSP_NAME,
+        PIPELINE_SOURCE_1, 2, s16le,
+        1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+# currently this dai is here as "virtual" capture backend
+W_DAI_IN(SSP, SPK_SSP_INDEX, SPK_SSP_NAME, s16le, 3, 0)
+
+# Capture pipeline 9 from demux on PCM 6 using max 2 channels of s32le.
+PIPELINE_PCM_ADD(sof/pipe-passthrough-capture-sched.m4,
+	9, 6, 2, s32le,
+	1000, 1, 0,
+	48000, 48000, 48000,
+	SCHEDULE_TIME_DOMAIN_TIMER,
+	PIPELINE_PLAYBACK_SCHED_COMP_1)
+
+# Connect demux to capture
+SectionGraph."PIPE_CAP" {
+	index "0"
+
+	lines [
+		# mux to capture
+		dapm(PIPELINE_SINK_9, PIPELINE_DEMUX_1)
+	]
+}
+
+# Connect virtual capture to dai
+SectionGraph."PIPE_CAP_VIRT" {
+	index "9"
+
+	lines [
+		# mux to capture
+		dapm(ECHO REF 9, SPK_REF_DAI_NAME)
+	]
+}')')
+
+#
+# Bind PCM with the pipeline
+#
+dnl PCM_PLAYBACK_ADD(name, pcm_id, playback, capture)
+PCM_DUPLEX_ADD(Headset, 1, PIPELINE_PCM_2, PIPELINE_PCM_3)
+PCM_PLAYBACK_ADD(HDMI1, 2, PIPELINE_PCM_5)
+PCM_PLAYBACK_ADD(HDMI2, 3, PIPELINE_PCM_6)
+PCM_PLAYBACK_ADD(HDMI3, 4, PIPELINE_PCM_7)
+PCM_PLAYBACK_ADD(HDMI4, 5, PIPELINE_PCM_8)
+ifdef(`NO_AMP',,`
+ifdef(`AMP_SSP',`',`fatal_error(note: Define AMP_SSP for speaker amp SSP Index)')
```

Comment:
it's better now, but could be even nicer if this error handling was done once at the top of the file.

ifdef(`NO_AMP',,`
￼ifdef(`AMP_SSP',`',`fatal_error(note: Define AMP_SSP for speaker amp SSP Index)'))

and later on you can simplify with 

ifdef(`NO_AMP',,`
ifdef(`SMART_AMP',,`

and assume that AMP_SSP is already defined.

Makes sense?

Reply:
@plbossart  Yes, agree! It makes simpler.